source/13_envs/evogym.rst

```diff
@@ -0,0 +1,226 @@
+
+2. Continuous action space according to the structure of the robot.
+
+3. Top-performing policies not only can learn how to control a robot, but they can also learn how to create a better robot design tailored for a specific task.
+
+Other
+======
+
+Robot and environment definition
+--------------------
+
+Evogym provides fixed tasks, but you can define your own robot structure and environment.
+For a concrete example you can refer to the following `json file
+<https://github.com/opendilab/DI-engine/blob/main/dizoo/evogym/envs/world_data/carry_bot.json>`__
+and to the official `documentation
+<https://evolutiongym.github.io/tutorials/new-env.html>`__.
+
+Lazy Initialization
+--------------------
+
+In order to support parallel operations such as environment vectorization, environment instances generally implement lazy initialization, that is, the \ ``__init__`` \ method does not initialize the real original environment instance, but only sets relevant parameters and configuration values. In the first call\ ``reset``\  method initializes the concrete original environment instance.
+
+Random Seed
+------------
+
+- There are two parts of the random seed in the environment that need to be set, one is the random seed of the original environment, and the other is the random seed of the random library used by various environment transformations (such as\ ``random``\ ，\ ``np.random``\ )
+
+- For the environment caller, just set these two seeds through the \ ``seed`` \ method of the environment, no need to care about the specific implementation details
+
+- Concrete implementation inside the environment: For the seed of the original environment, set before calling the\ ``reset``\ method of the environment, the concrete original environment\ ``reset``\ 
+
+- Concrete implementation inside the environment: For random library seeds, the value is set directly in the \ ``seed`` \ method of the environment
+
+The Difference between Training and Testing Environments
+---------------------------------------------------------
+
+- The training environment uses a dynamic random seed, that is, the random seed of each episode is different, and is generated by a random number generator, but the seed of this random number generator is fixed by the \ ``seed`` \ method of the environment ;The test environment uses a static random seed, that is, the random seed of each episode is the same, specified by the \ ``seed`` \ method.
+
+Store Video
+------------
+
+After the environment is created, but before reset, call the \ ``enable_save_replay`` \ method, specifying the path to save the game replay. The environment will automatically save the local video files after each episode ends. (The default call \ ``gym.wrappers.RecordVideo`` \ implementation ), the code shown below will run an environment episode and save the result of this episode in a folder\ ``./video/``\ :
+
+.. code:: python
+
+    import gym
```

Comment:
replace with:

import time
import gym
from evogym import sample_robot
# import envs from the envs folder and register them
import evogym.envs

if __name__ == '__main__':
    # create a random robot
    body, connections = sample_robot((5, 5))
    env = gym.make('Walker-v0', body=body)

    if gym.version.VERSION > '0.22.0':
        env.metadata.update({'render_modes': ["rgb_array"]})
    else:
        env.metadata.update({'render.modes': ["rgb_array"]})
    env = gym.wrappers.RecordVideo(
        env,
        video_folder="./video",
        episode_trigger=lambda episode_id: True,
        name_prefix='rl-video-{}'.format(time.time())
    )

    env.reset()
    # step the environment for 100 iterations
    for i in range(100):
        action = env.action_space.sample()
        ob, reward, done, info = env.step(action)
        x = env.render()
        if done:
            env.reset()
    env.close()